api/src/service/cache_snapshot.ts

```diff
@@ -0,0 +1,413 @@
+  }
+  let parsedEvents = parseBusinessEvents(items, streamName);
+  const businessEvents = parsedEvents.filter(Result.isOk);
+
+  const sourcedData = sourceFromSnapshot(ctx, businessEvents, false, data);
+  if (ordering && eventType === SUBPROJECT_SNAPSHOT_EVENT_KEY) {
+    if (Result.isOk(sourcedData)) {
+      sourcedData.workflowitemOrdering = ordering;
+    }
+  }
+  let publishEvent: Result.Type<BusinessEvent>;
+  if (eventType === WORKFLOWITEM_SNAPSHOT_EVENT_KEY) {
+    publishEvent = createEvent(
+      ctx.source,
+      creatingUser.id,
+      streamName,
+      sourcedData.subprojectId,
+      sourcedData,
+    );
+  } else {
+    publishEvent = createEvent(ctx.source, creatingUser.id, sourcedData);
+  }
+
+  return { canPublish: true, eventData: publishEvent };
+}
+
+function getSourceInfo(
+  key: string,
+  eventType: string,
+): { searchKey: string; sourceFromSnapshot: Function; parseFromSnapshot: Function } {
+  let searchKey;
+  let sourceFromSnapshot: Function;
+  let parseFromSnapshot: Function;
+
+  switch (eventType) {
+    case PROJECT_SNAPSHOT_EVENT_KEY:
+      searchKey = SNAPSHOT_KEY;
+      sourceFromSnapshot = ProjectEventSourcing.sourceProjectFromSnapshot;
+      parseFromSnapshot = ProjectEventSourcing.parseProjectFromSnapshot;
+      break;
+    case SUBPROJECT_SNAPSHOT_EVENT_KEY:
+      searchKey = key + "_" + SNAPSHOT_KEY;
+      sourceFromSnapshot = SubprojectEventSourcing.sourceSubprojectFromSnapshot;
+      parseFromSnapshot = SubprojectEventSourcing.parseSubprojectFromSnapshot;
+      break;
+    case WORKFLOWITEM_SNAPSHOT_EVENT_KEY:
+      searchKey = key + "_" + SNAPSHOT_KEY;
+      sourceFromSnapshot = WorkflowitemEventSourcing.sourceWorkflowitemFromSnapshot;
+      parseFromSnapshot = WorkflowitemEventSourcing.parseWorkflowitemFromSnapshot;
+      break;
+    default:
+      // eslint-disable-next-line @typescript-eslint/explicit-function-return-type, @typescript-eslint/no-empty-function
+      return { searchKey: "", sourceFromSnapshot: () => {}, parseFromSnapshot: () => {} };
+  }
+  return {
+    searchKey: searchKey,
+    sourceFromSnapshot: sourceFromSnapshot,
+    parseFromSnapshot: parseFromSnapshot,
+  };
+}
+
+async function getItemsInfo(
+  conn: ConnToken,
+  ctx: Ctx,
+  streamName: string,
+  key: string,
+  searchKey: string,
+): Promise<Result.Type<{ items: Item[]; snapshotIndex: number }>> {
+  const rpcClient = conn.multichainClient.getRpcClient();
+  let items: Item[] = [];
+  try {
+    items = await rpcClient.invoke("liststreamkeyitems", streamName, key, false, MAX_ITEM_COUNT);
+    if (items.length == 0) {
+      return new VError("Data Not Found");
+    }
+  } catch (e) {
+    return new VError("Data Not Found");
+  }
+
+  // Traverse the list reverse to get latest snapshot
+  let snapshotIndex = -1;
+  for (let i = items.length - 1; i >= 0; i--) {
+    let item = items[i];
+    if (item.keys.includes(searchKey)) {
+      snapshotIndex = i;
+      break;
+    }
+  }
```

Comment:
Counter-opinion: The snapshot events are usually near the end of the items array. That's why here we start looking from the end of the array to the beginning in reverse order, because for very long list of items this would be much more efficient and faster to find the snapshot. findIndex iterates in an ascending-index order. If we want to use findIndex, we have to in-place reverse the items array (redundant inefficiency for large items array) which messes up the event order, or we copy the array and reverse a copy of it but copying a whole array if the array size is big is not memory friendly.